File: src/lib/git.ts
import axios from 'axios';

export async function repositoryExists(owner: string, repo: string, token: string): Promise<boolean> {
    console.log(`[git.ts] Checking if repository exists: ${owner}/${repo}`);
    try {
        await axios.get(
            `https://api.github.com/repos/${owner}/${repo}`,
            {
                headers: { Authorization: `token ${token}` }
            }
        );
        return true;
    } catch (error: any) {
        if (error.response?.status === 404) {
            console.log(`[git.ts] Repository ${owner}/${repo} not found (404).`);
            return false;
        }
        console.error(`[git.ts] Error checking repository existence: ${error.message} Status: ${error.response?.status}`);
        throw error;
    }
}

/**
 * Configuration options for fetching repository content
 */
interface FetchRepositoryConfig {
    /** File extensions to include (if not specified, includes all files) */
    includeExtensions?: string[];
    /** File paths or patterns to exclude */
    excludePatterns?: string[];
    /** Maximum file size in bytes (default: 1MB) */
    maxFileSize?: number;
    /** Whether to include binary files (default: false) */
    includeBinary?: boolean;
    /** Custom function to filter files */
    customFilter?: (path: string, content: string) => boolean;
}

/**
 * Fetch all files from a repository with their full paths and content
 * @param owner - Repository owner
 * @param repo - Repository name
 * @param token - GitHub personal access token
 * @param branch - Branch name (default: 'main')
 * @param config - Configuration options for filtering files
 * @returns Array of files with path and content
 */
export async function fetchAllRepositoryFiles(
    owner: string,
    repo: string,
    token: string,
    branch: string = 'main',
    config?: FetchRepositoryConfig
): Promise<{ path: string; content: string }[]> {
    console.log(`[git.ts] Fetching all files from ${owner}/${repo}/${branch}`);

    // Set default configuration values
    const {
        includeExtensions,
        excludePatterns,
        maxFileSize = 1024 * 1024, // 1MB default
        includeBinary = false,
        customFilter
    } = config || {};

    try {
        // Get the repository tree recursively
        const res = await axios.get(
            `https://api.github.com/repos/${owner}/${repo}/git/trees/${branch}?recursive=1`,
            {
                headers: { Authorization: `token ${token}` }
            }
        );

        const tree = res.data.tree;
        const files: { path: string; content: string }[] = [];
        let skippedCount = 0;

        // Process each item in the tree
        for (const item of tree) {
            // Only process files (not directories or submodules)
            if (item.type === 'blob') {
                try {
                    // Check if file should be excluded based on patterns
                    if (excludePatterns && excludePatterns.some(pattern => {
                        // Convert glob patterns to regex
                        const regex = new RegExp(
                            pattern.replace(/\*/g, '.*').replace(/\?/g, '.')
                        );
                        return regex.test(item.path);
                    })) {
                        console.log(`[git.ts] Skipping excluded file: ${item.path}`);
                        skippedCount++;
                        continue;
                    }

                    // Check file extension if includeExtensions is specified
                    if (includeExtensions && includeExtensions.length > 0) {
                        const fileExtension = item.path.split('.').pop()?.toLowerCase();
                        if (!fileExtension || !includeExtensions.includes(fileExtension)) {
                            console.log(`[git.ts] Skipping file with non-matching extension: ${item.path}`);
                            skippedCount++;
                            continue;
                        }
                    }

                    // Fetch the content of each file
                    const contentRes = await axios.get(
                        item.url,
                        {
                            headers: { Authorization: `token ${token}` }
                        }
                    );

                    // Check file size
                    if (contentRes.data.size > maxFileSize) {
                        console.log(`[git.ts] Skipping large file (${contentRes.data.size} bytes): ${item.path}`);
                        skippedCount++;
                        continue;
                    }

                    // Decode base64 content
                    let content: string;
                    try {
                        content = Buffer.from(contentRes.data.content, 'base64').toString('utf-8');
                    } catch (error) {
                        if (!includeBinary) {
                            console.log(`[git.ts] Skipping binary file: ${item.path}`);
                            skippedCount++;
                            continue;
                        }
                        // For binary files, keep the base64 content
                        content = contentRes.data.content;
                    }

                    // Apply custom filter if provided
                    if (customFilter && !customFilter(item.path, content)) {
                        console.log(`[git.ts] Skipping file due to custom filter: ${item.path}`);
                        skippedCount++;
                        continue;
                    }

                    files.push({
                        path: item.path,
                        content
                    });

                    console.log(`[git.ts] Fetched file: ${item.path}`);
                } catch (error: any) {
                    console.error(`[git.ts] Error fetching file ${item.path}:`, error.message);
                    skippedCount++;
                    // Continue with other files even if one fails
                }
            }
        }

        console.log(`[git.ts] Successfully fetched ${files.length} files from ${owner}/${repo}. Skipped ${skippedCount} files.`);
        return files;
    } catch (error: any) {
        console.error(`[git.ts] Error fetching repository files: ${error.message}`);
        if (error.response?.status === 404) {
            throw new Error(`Repository ${owner}/${repo} or branch ${branch} not found`);
        } else if (error.response?.status === 409) {
            throw new Error(`Repository ${owner}/${repo} is empty`);
        }
        throw error;
    }
}

/**
 * Convenience function to fetch only source code files
 * @param owner - Repository owner
 * @param repo - Repository name
 * @param token - GitHub personal access token
 * @param branch - Branch name (default: 'main')
 * @returns Array of source code files with path and content
 */
export async function fetchSourceCodeFiles(
    owner: string,
    repo: string,
    token: string,
    branch: string = 'main'
): Promise<{ path: string; content: string }[]> {
    const sourceExtensions = [
        'js', 'jsx', 'ts', 'tsx', 'vue', 'svelte',
        'py', 'java', 'cpp', 'c', 'h', 'hpp',
        'cs', 'php', 'rb', 'go', 'rs', 'swift',
        'kt', 'scala', 'clj', 'hs', 'ml', 'sh',
        'sql', 'html', 'css', 'scss', 'less', 'sass',
        'json', 'xml', 'yaml', 'yml', 'toml', 'ini',
        'md', 'txt', 'dockerfile', 'gitignore'
    ];

    return fetchAllRepositoryFiles(owner, repo, token, branch, {
        includeExtensions: sourceExtensions,
        excludePatterns: [
            'node_modules/*',
            'dist/*',
            'build/*',
            '*.min.js',
            '*.min.css',
            'package-lock.json',
            'yarn.lock'
        ]
    });
}



export async function getBranchSHA(owner: string, repo: string, branch: string, token: string) {
    const res = await axios.get(
        `https://api.github.com/repos/${owner}/${repo}/git/refs/heads/${branch}`,
        {
            headers: { Authorization: `token ${token}` }
        }
    );

    return res.data.object.sha;
}

export async function getCommitTreeSHA(owner: string, repo: string, commitSHA: string, token: string) {
    const res = await axios.get(
        `https://api.github.com/repos/${owner}/${repo}/git/commits/${commitSHA}`,
        {
            headers: { Authorization: `token ${token}` }
        }
    );
    return res.data.tree.sha;
}

export async function commitFilesToBranch(
    owner: string,
    repo: string,
    branch: string,
    files: { path: string; content: string }[],
    message: string,
    token: string
) {
    // 1. Get Tip
    const baseSHA = await getBranchSHA(owner, repo, branch, token);

    // 2. Get Tree
    const baseTreeSHA = await getCommitTreeSHA(owner, repo, baseSHA, token);

    // 3. Create New Tree
    const newTreeSHA = await createTree(owner, repo, baseTreeSHA, files, token);

    // 4. Create Commit
    const commitSHA = await createCommit(owner, repo, baseSHA, newTreeSHA, message, token);

    // 5. Update Ref
    await updateBranch(owner, repo, branch, commitSHA, token);

    return commitSHA;
}

export async function createBlob(owner: string, repo: string, content: string, token: string) {
    const res = await axios.post(
        `https://api.github.com/repos/${owner}/${repo}/git/blobs`,
        {
            content,
            encoding: "utf-8"
        },
        {
            headers: { Authorization: `token ${token}` }
        }
    );

    return res.data.sha;
}

export async function createTree(owner: string, repo: string, baseTreeSHA: string, files: { path: string; content: string }[], token: string) {
    const tree = [];

    for (const file of files) {
        const blobSHA = await createBlob(owner, repo, file.content, token);

        tree.push({
            path: file.path,
            mode: "100644",
            type: "blob",
            sha: blobSHA
        });
    }

    const res = await axios.post(
        `https://api.github.com/repos/${owner}/${repo}/git/trees`,
        {
            base_tree: baseTreeSHA,
            tree
        },
        {
            headers: { Authorization: `token ${token}` }
        }
    );

    return res.data.sha;
}

export async function createCommit(owner: string, repo: string, parentSHA: string, treeSHA: string, message: string, token: string) {
    const res = await axios.post(
        `https://api.github.com/repos/${owner}/${repo}/git/commits`,
        {
            message,
            parents: [parentSHA],
            tree: treeSHA
        },
        { headers: { Authorization: `token ${token}` } }
    );

    return res.data.sha;
}

export async function updateBranch(owner: string, repo: string, branch: string, commitSHA: string, token: string) {
    await axios.patch(
        `https://api.github.com/repos/${owner}/${repo}/git/refs/heads/${branch}`,
        {
            sha: commitSHA,
            force: true
        },
        {
            headers: { Authorization: `token ${token}` }
        }
    );
}

export async function createBranch(owner: string, repo: string, newBranch: string, baseSHA: string, token: string) {
    console.log(`[git.ts] Creating branch ${newBranch} from ${baseSHA} in ${owner}/${repo}`);
    try {
        await axios.post(
            `https://api.github.com/repos/${owner}/${repo}/git/refs`,
            {
                ref: `refs/heads/${newBranch}`,
                sha: baseSHA
            },
            {
                headers: { Authorization: `token ${token}` }
            }
        );
    } catch (error: any) {
        // If branch already exists (422), ignore the error
        if (error.response?.status === 422) {
            console.log(`Branch ${newBranch} already exists. Skipping creation.`);
            return;
        }
        throw error;
    }
}

/**
 * Create a new GitHub repository
 * @param name - Repository name
 * @param isPrivate - Whether the repository should be private (default: true)
 * @param description - Repository description (optional)
 * @param token - GitHub personal access token
 * @param org - Organization name (optional, creates under user if not provided)
 * @returns Repository data including full_name, html_url, clone_url, etc.
 */
export async function createRepository(
    name: string,
    isPrivate: boolean,
    description: string | undefined,
    token: string,
    org?: string
) {
    const endpoint = org
        ? `https://api.github.com/orgs/${org}/repos`
        : 'https://api.github.com/user/repos';


    try {

        const res = await axios.post(
            endpoint,
            {
                name,
                description,
                private: isPrivate,
                auto_init: true, // Initialize with README to enable Git Data API
            },
            {
                headers: { Authorization: `token ${token}` }
            }
        );

        return {
            name: res.data.name,
            fullName: res.data.full_name,
            htmlUrl: res.data.html_url,
            cloneUrl: res.data.clone_url,
            owner: res.data.owner.login,
        };
    } catch (error) {
        console.log(error);
        throw error;
    }
}

/**
 * Check if a repository exists
 * @param owner - Repository owner
 * @param repo - Repository name
 * @param token - GitHub personal access token
 * @returns True if repository exists, false otherwise
 */


/**
 * Verify token permissions and owner
 * @param token - GitHub personal access token
 * @returns Object containing login and scopes
 */
export async function verifyToken(token: string) {
    console.log(`[git.ts] Verifying token permissions...`);
    try {
        const res = await axios.get('https://api.github.com/user', {
            headers: { Authorization: `token ${token}` }
        });

        const scopes = res.headers['x-oauth-scopes'] || '';
        console.log(`[git.ts] Token verified. User: ${res.data.login}, Scopes: ${scopes}`);

        return {
            login: res.data.login,
            scopes: scopes.split(',').map((s: string) => s.trim())
        };
    } catch (error: any) {
        console.error(`[git.ts] Token verification failed: ${error.message}`);
        if (error.response) {
            console.error(`[git.ts] Status: ${error.response.status}`);
            console.error(`[git.ts] Body: ${JSON.stringify(error.response.data)}`);
        }
        throw new Error(`GitHub token verification failed: ${error.message}`);
    }
}

export async function getGithubOwner(): Promise<string> {    
    const token = process.env.GITHUB_TOKEN;

    if(!token) {
        throw new Error('GITHUB_TOKEN is not set');
    }

    try {
        const { login } = await verifyToken(token);
        return login;
    } catch (error: any) {
        console.error(`[git.ts] Error fetching GitHub owner: ${error.message}`);
        throw error;
    }
}

export async function pushFilesAsCommit({
    owner,
    repo,
    baseBranch,
    newBranch,
    files,
    commitMessage,
    token,
    repoDescription
}: {
    owner: string;
    repo: string;
    baseBranch: string;
    newBranch: string;
    files: { path: string; content: string }[];
    commitMessage: string;
    token: string;
    repoDescription?: string;
}) {
    console.log(`[git.ts] pushFilesAsCommit called for ${owner}/${repo}`);
    const tokenMasked = token ? `${token.substring(0, 7)}...${token.substring(token.length - 4)}` : 'undefined';
    console.log(`[git.ts] Token present: ${!!token}, length: ${token?.length}, value: ${tokenMasked}`);

    try {
        // 0. Verify token first
        const tokenInfo = await verifyToken(token);

        // Check if token owner matches requested owner (unless it's an org, which is harder to check without more calls)
        // But logging it is enough for debugging
        if (tokenInfo.login.toLowerCase() !== owner.toLowerCase()) {
            console.warn(`[git.ts] WARNING: Token owner (${tokenInfo.login}) does not match requested owner (${owner}). This might be intended if pushing to an organization.`);
        }

        // Check for required scopes
        const requiredScopes = ['repo']; // 'repo' covers everything for private repos. 'public_repo' for public.
        const hasRepoScope = tokenInfo.scopes.includes('repo');
        const hasPublicRepoScope = tokenInfo.scopes.includes('public_repo');

        if (!hasRepoScope && !hasPublicRepoScope) {
            console.warn(`[git.ts] WARNING: Token seems to be missing 'repo' or 'public_repo' scope. Scopes found: ${tokenInfo.scopes.join(', ')}`);
        }

        // 1. Check if repository exists, create if not
        const exists = await repositoryExists(owner, repo, token);

        if (!exists) {
            console.log(`Repository ${owner}/${repo} does not exist. Creating...`);
            const newRepo = await createRepository(
                repo,
                false, // public repository to match token scope
                repoDescription || `Generated repository from v0`,
                token
            );
            console.log(`Repository created: ${newRepo.htmlUrl}`);

            // Wait a moment for GitHub to fully initialize the repository
            await new Promise(resolve => setTimeout(resolve, 2000));
        }

        // 2. Get base branch SHA
        console.log(`[git.ts] Getting SHA for branch ${baseBranch}...`);
        const baseSHA = await getBranchSHA(owner, repo, baseBranch, token);

        // If creating a new branch
        if (newBranch) {
            await createBranch(owner, repo, newBranch, baseSHA, token);
            baseBranch = newBranch;
        }

        // 3. Get tree SHA of base branch
        console.log(`[git.ts] Getting tree SHA for commit ${baseSHA}...`);
        const res = await axios.get(
            `https://api.github.com/repos/${owner}/${repo}/git/commits/${baseSHA}`,
            {
                headers: { Authorization: `token ${token}` }
            }
        );
        const baseTreeSHA = res.data.tree.sha;

        // Validate and sanitize files before creating tree
        console.log(`[git.ts] Validating ${files.length} files before creating tree...`);
        const validatedFiles = files
            .filter(file => {
                // Filter out files with empty or invalid content
                if (!file.content || file.content.trim().length === 0) {
                    console.warn(`[git.ts] Skipping empty file: ${file.path}`);
                    return false;
                }

                // Filter out files with invalid paths
                if (!file.path || file.path.trim().length === 0) {
                    console.warn(`[git.ts] Skipping file with empty path`);
                    return false;
                }

                // GitHub doesn't allow certain characters in file paths
                const invalidChars = /[<>:"|?*\x00-\x1F]/;
                if (invalidChars.test(file.path)) {
                    console.warn(`[git.ts] Skipping file with invalid characters in path: ${file.path}`);
                    return false;
                }

                return true;
            })
            .map(file => ({
                // Normalize path separators to forward slashes
                path: file.path.replace(/\\/g, '/').trim(),
                content: file.content
            }));

        if (validatedFiles.length === 0) {
            throw new Error('No valid files to commit after validation');
        }

        console.log(`[git.ts] ${validatedFiles.length} files passed validation`);
        if (validatedFiles.length < files.length) {
            console.warn(`[git.ts] ${files.length - validatedFiles.length} files were filtered out`);
        }

        // Log file paths for debugging
        console.log(`[git.ts] File paths to be committed:`, validatedFiles.map(f => f.path).join(', '));

        // 4. Create new tree
        console.log(`[git.ts] Creating new tree with ${validatedFiles.length} files...`);
        const newTreeSHA = await createTree(
            owner,
            repo,
            baseTreeSHA,
            validatedFiles,
            token
        );

        // 5. Create commit
        console.log(`[git.ts] Creating commit...`);
        const commitSHA = await createCommit(
            owner,
            repo,
            baseSHA,
            newTreeSHA,
            commitMessage,
            token
        );

        // 6. Update branch
        console.log(`[git.ts] Updating branch ${baseBranch} to ${commitSHA}...`);
        await updateBranch(owner, repo, baseBranch, commitSHA, token);

        return {
            commitSHA,
            repoUrl: `https://github.com/${owner}/${repo}`,
            branchUrl: `https://github.com/${owner}/${repo}/tree/${baseBranch}`,
        };
    } catch (error: any) {
        console.error("Error in pushFilesAsCommit:", error.response?.data || error.message);
        console.error("Failed Request URL:", error.config?.url);
        console.error("Failed Request Method:", error.config?.method);
        console.error("Status Code:", error.response?.status);

        // Log full response for 422 errors
        if (error.response?.status === 422) {
            console.error("422 Validation Error Details:");
            console.error("Response Body:", JSON.stringify(error.response?.data, null, 2));
            console.error("Request Body:", JSON.stringify(error.config?.data, null, 2));

            // Create user-friendly error message
            const errorMsg = error.response?.data?.message || 'Validation failed';
            const errors = error.response?.data?.errors || [];
            let detailedError = `GitHub API validation error: ${errorMsg}`;

            if (errors.length > 0) {
                detailedError += '\nDetails: ' + errors.map((e: any) =>
                    `${e.resource || 'unknown'}: ${e.message || e.code || JSON.stringify(e)}`
                ).join(', ');
            }

            throw new Error(detailedError);
        }

        // Provide helpful message for common errors
        if (error.response?.status === 409) {
            console.error("409 Conflict - This usually means the repository or branch already exists.");
            console.error("Repository:", `${owner}/${repo}`);
        }

        throw error;
    }
}

File: agent/utils/git.py
import asyncio
import httpx
from utils.settings import settings
from utils.vector_store import upsert_document

async def get_github_owner() -> str:
    """Fetch the authenticated user's login (owner name) from GitHub."""
    token = settings.github_token
    async with httpx.AsyncClient() as client:
        response = await client.get(
            "https://api.github.com/user",
            headers={"Authorization": f"token {token}"}
        )
        if response.status_code == 200:
            return response.json()["login"]
        else:
            raise Exception(f"Failed to fetch user. Status: {response.status_code}")

async def sync_repository_to_vector_store(owner: str, repo: str) -> str:
    """
    Fetches all text files from the repository, chunks them, and syncs to Upstash Vector Store.
    Returns a summary string of the operation.
    """
    from utils.chunking import chunk_text, should_skip_file
    
    token = settings.github_token
    headers = {"Authorization": f"token {token}"}
    namespace = f"{owner}-{repo}"
    
    # 1. Get Tree Recursive
    async with httpx.AsyncClient() as client:
        # Get default branch sha first to be safe, or just recursive tree from 'main'
        tree_url = f"https://api.github.com/repos/{owner}/{repo}/git/trees/main?recursive=1"
        response = await client.get(tree_url, headers=headers)
        
        if response.status_code != 200:
             # Try master if main fails
             tree_url = f"https://api.github.com/repos/{owner}/{repo}/git/trees/master?recursive=1"
             response = await client.get(tree_url, headers=headers)
             if response.status_code != 200:
                # Add better error logging
                error_msg = f"Failed to fetch repository tree for {owner}/{repo}. Status: {response.status_code}"
                try:
                    error_detail = response.json()
                    error_msg += f", Detail: {error_detail.get('message', 'Unknown error')}"
                except:
                    pass
                print(f"ERROR: {error_msg}")
                raise Exception(error_msg)

        tree_data = response.json()
        
        # 2. Filter & Fetch Files with Chunking
        synced_count = 0
        chunk_count = 0
        skipped_count = 0
        
        for item in tree_data.get("tree", []):
            if item["type"] == "blob":
                path = item["path"]
                
                # Fetch content first to check if we should skip
                blob_url = item["url"] # API url for blob
                blob_resp = await client.get(blob_url, headers=headers)
                if blob_resp.status_code == 200:
                    import base64
                    try:
                        content_b64 = blob_resp.json()["content"]
                        content = base64.b64decode(content_b64).decode("utf-8")
                        
                        # Check if we should skip this file
                        if should_skip_file(path, content):
                            skipped_count += 1
                            continue
                        
                        # Chunk the content
                        chunks = chunk_text(content)
                        
                        # Upsert each chunk to Vector Store
                        for i, chunk in enumerate(chunks):
                            chunk_id = f"{path}-chunk-{i}"
                            metadata = {
                                "repository": f"{owner}/{repo}", 
                                "path": path,
                                "chunk_index": i,
                                "total_chunks": len(chunks)
                            }
                            await asyncio.to_thread(
                                upsert_document,
                                doc_id=chunk_id, 
                                content=chunk, 
                                metadata=metadata, 
                                namespace=namespace
                            )
                            chunk_count += 1
                        
                        synced_count += 1
                    except Exception as e:
                        print(f"Failed to parse/sync {path}: {e}")
                        skipped_count += 1
                        
        return f"Synced {synced_count} files ({chunk_count} chunks) to vector store. Skipped {skipped_count}."


async def sync_changed_files_to_vector(owner: str, repo: str, files: list[dict]) -> str:
    """
    Sync only specific changed files to vector store (incremental update).
    
    Args:
        owner: Repository owner
        repo: Repository name
        files: List of file dicts with 'path' and 'content' keys
    
    Returns:
        Summary string
    """
    from utils.chunking import chunk_text, should_skip_file
    
    namespace = f"{owner}-{repo}"
    synced_count = 0
    chunk_count = 0
    
    for file in files:
        path = file.get("path") or file.get("file_name")
        content = file.get("content", "")
        
        # Skip if needed
        if should_skip_file(path, content):
            continue
        
        # Chunk and upsert
        chunks = chunk_text(content)
        for i, chunk in enumerate(chunks):
            chunk_id = f"{path}-chunk-{i}"
            metadata = {
                "repository": f"{owner}/{repo}",
                "path": path,
                "chunk_index": i,
                "total_chunks": len(chunks)
            }
            upsert_document(
                doc_id=chunk_id,
                content=chunk,
                metadata=metadata,
                namespace=namespace
            )
            chunk_count += 1
        
        synced_count += 1
    
    return f"Incrementally synced {synced_count} files ({chunk_count} chunks)"


async def repositoryExists(repository_name: str) -> bool:
    try:
        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"https://api.github.com/repos/{repository_name}",
                headers={"Authorization": f"token {settings.github_token}"}
            )
            if response.status_code == 200:
                return True
            elif response.status_code == 404:
                return False
            else:
                raise Exception(f"Error checking repository existence: {response.status_code}")
    except Exception as e:
        raise e

async def get_branch_sha(owner: str, repo: str, branch: str, token: str, client: httpx.AsyncClient) -> str:
    response = await client.get(
        f"https://api.github.com/repos/{owner}/{repo}/git/refs/heads/{branch}",
        headers={"Authorization": f"token {token}"}
    )
    if response.status_code != 200:
        raise Exception(f"Failed to get branch SHA for {branch}: {response.status_code} - {response.text}")
    return response.json()["object"]["sha"]

async def get_commit_tree_sha(owner: str, repo: str, commit_sha: str, token: str, client: httpx.AsyncClient) -> str:
    response = await client.get(
        f"https://api.github.com/repos/{owner}/{repo}/git/commits/{commit_sha}",
        headers={"Authorization": f"token {token}"}
    )
    if response.status_code != 200:
        raise Exception(f"Failed to get commit tree SHA: {response.status_code} - {response.text}")
    return response.json()["tree"]["sha"]

async def create_blob(owner: str, repo: str, content: str, token: str, client: httpx.AsyncClient) -> str:
    response = await client.post(
        f"https://api.github.com/repos/{owner}/{repo}/git/blobs",
        headers={"Authorization": f"token {token}"},
        json={"content": content, "encoding": "utf-8"}
    )
    if response.status_code != 201:
         raise Exception(f"Failed to create blob: {response.status_code} - {response.text}")
    return response.json()["sha"]

async def create_tree(owner: str, repo: str, base_tree_sha: str, files: list, token: str, client: httpx.AsyncClient) -> str:
    tree_items = []
    for file in files:
        blob_sha = await create_blob(owner, repo, file["content"], token, client)
        # Handle both 'path' (from development workflow) and 'file_name' (from initialize_project)
        file_path = file.get("path") or file.get("file_name")
        tree_items.append({
            "path": file_path,
            "mode": "100644",
            "type": "blob",
            "sha": blob_sha
        })
    
    response = await client.post(
        f"https://api.github.com/repos/{owner}/{repo}/git/trees",
        headers={"Authorization": f"token {token}"},
        json={"base_tree": base_tree_sha, "tree": tree_items}
    )
    if response.status_code != 201:
        raise Exception(f"Failed to create tree: {response.status_code} - {response.text}")
    return response.json()["sha"]

async def create_commit(owner: str, repo: str, parent_sha: str, tree_sha: str, message: str, token: str, client: httpx.AsyncClient) -> str:
    response = await client.post(
        f"https://api.github.com/repos/{owner}/{repo}/git/commits",
        headers={"Authorization": f"token {token}"},
        json={
            "message": message,
            "parents": [parent_sha],
            "tree": tree_sha
        },
    )
    if response.status_code != 201:
        raise Exception(f"Failed to create commit: {response.status_code} - {response.text}")
    return response.json()["sha"]

async def update_branch(owner: str, repo: str, branch: str, commit_sha: str, token: str, client: httpx.AsyncClient):
    response = await client.patch(
        f"https://api.github.com/repos/{owner}/{repo}/git/refs/heads/{branch}",
        headers={"Authorization": f"token {token}"},
        json={"sha": commit_sha, "force": True}
    )
    if response.status_code != 200:
        raise Exception(f"Failed to update branch: {response.status_code} - {response.text}")

async def createRepository(repo_name: str, description: str, private: bool = False) -> str:
    """Creates a new GitHub repository."""
    token = settings.github_token
    try:
        async with httpx.AsyncClient() as client:
            response = await client.post(
                "https://api.github.com/user/repos",
                headers={
                    "Authorization": f"token {token}",
                    "Accept": "application/vnd.github.v3+json"
                },
                json={
                    "name": repo_name,
                    "description": description,
                    "private": private,
                    "auto_init": True 
                }
            )
            
            if response.status_code in (200, 201):
                return response.json()["html_url"]
            else:
                raise Exception(f"Failed to create repository: {response.status_code} - {response.text}")
    except Exception as e:
        raise Exception(f"Error creating repository: {str(e)}")

async def push_files(owner: str, repo_name: str, files: list, commit_message: str):
    """
    Orchestrates the git push mechanism using low-level Git Data API.
    Replicates logic from src/lib/git.ts.
    """
    token = settings.github_token
    
    async with httpx.AsyncClient() as client:
        # 1. Get Base SHA (assuming main)
        # Note: New repos created with auto_init will have a 'main' branch
        base_sha = await get_branch_sha(owner, repo_name, "main", token, client)
        
        # 2. Get Base Tree SHA
        base_tree_sha = await get_commit_tree_sha(owner, repo_name, base_sha, token, client)
        
        # 3. Create New Tree (Blobs + Tree)
        new_tree_sha = await create_tree(owner, repo_name, base_tree_sha, files, token, client)
        
        # 4. Create Commit
        new_commit_sha = await create_commit(owner, repo_name, base_sha, new_tree_sha, commit_message, token, client)
        
        # 5. Update Branch Ref
        await update_branch(owner, repo_name, "main", new_commit_sha, token, client)
        
        return f"https://github.com/{owner}/{repo_name}"


async def fetch_golden_context(owner: str, repo: str) -> str:
    """
    Fetch golden context files (brief, specs, etc) from GitHub.
    Returns a consolidated string or empty string if no files found.
    """
    token = settings.github_token
    headers = {"Authorization": f"token {token}"}
    
    context_files = [
        "project_brief.md",
        "technical_spec.md",
        "implementation_plan.md",
        "coding_guidelines.md",
        "README.md"
    ]
    
    golden_context_parts = []
    async with httpx.AsyncClient() as client:
        for filename in context_files:
            try:
                url = f"https://api.github.com/repos/{owner}/{repo}/contents/{filename}"
                response = await client.get(url, headers=headers)
                
                if response.status_code == 200:
                    import base64
                    content_b64 = response.json()["content"]
                    content = base64.b64decode(content_b64).decode("utf-8")
                    golden_context_parts.append(f"# {filename}\n\n{content}")
            except Exception:
                # Silently skip missing/failed files
                pass
    
    if not golden_context_parts:
        return ""
        
    return "\n\n---\n\n".join(golden_context_parts)